src/price/ProgressivePrice.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types=1);
+
+namespace hiqdev\php\billing\price;
+
+use hiqdev\php\billing\plan\PlanInterface;
+use hiqdev\php\billing\target\TargetInterface;
+use hiqdev\php\billing\type\TypeInterface;
+use hiqdev\php\units\Quantity;
+use hiqdev\php\units\QuantityInterface;
+use Money\Currency;
+use Money\Money;
+
+class ProgressivePrice extends AbstractPrice
+{
+    /* @var ProgressivePriceConditionDto[] $thresholds */
+    protected array $thresholds;
+    /**
+     * @var QuantityInterface prepaid quantity also implies Unit
+     * XXX cannot be null cause Unit is required
+     */
+    protected $prepaid;
+
+    public function __construct(
+        $id,
+        TypeInterface $type,
+        TargetInterface $target,
+        QuantityInterface $prepaid,
+        array $thresholds,
+        ?PlanInterface $plan = null
+    ) {
+        parent::__construct($id, $type, $target, $plan);
+        $this->prepaid = $prepaid;
+        $this->thresholds = $thresholds;
+    }
+
+    public function getPrepaid()
+    {
+        return $this->prepaid;
+    }
+
+    public function getThresholds(): array
+    {
+        return $this->thresholds;
+    }
+
+    private function prepareThresholds(): void
+    {
+        usort($this->thresholds, function($a, $b)
+            {
+                return $b->value <=> $a->value;
+            }
+        );
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function calculateUsage(QuantityInterface $quantity): ?QuantityInterface
+    {
+        $usage = $quantity->convert($this->prepaid->getUnit())->subtract($this->prepaid);
+
+        if ($usage->isPositive()) {
+            return $usage;
+        }
+
+        return Quantity::create($this->prepaid->getUnit()->getName(), 0);
```

Comment:
Matches **SinglePrice**.

Why did you prefer to extend AbstractPrice?

Reply:
I extended the abstract class because it had base methods for plan, target, etc.

Reply:
The same methods are in **SinglePrice**, as it extends **AbstractPrice**. Maybe we can extend it not to duplicate method implementations?